src/main/java/org/antlr/jetbrains/adapter/lexer/PsiElementTypeFactory.java

```diff
@@ -92,11 +100,16 @@ public TokenIElementType getEofElementType() {
     }
 
     @NotNull
-    private List<RuleIElementType> createRuleIElementTypes(Language language, String[] ruleNames) {
+    private List<RuleIElementType> createRuleIElementTypes(Language language, String[] ruleNames,
+            Collection<RuleIElementType> customRuleElementTypes) {
         List<RuleIElementType> result;
-        RuleIElementType[] elementTypes = new RuleIElementType[ruleNames.length];
+        RuleIElementType[] elementTypes = new RuleIElementTypeImpl[ruleNames.length];
         for (int i = 0; i < ruleNames.length; i++) {
-            elementTypes[i] = new RuleIElementType(i, ruleNames[i], language);
+            elementTypes[i] = new RuleIElementTypeImpl(i, ruleNames[i], language);
+        }
+        for (RuleIElementType customType : customRuleElementTypes) {
+            assert customType instanceof IElementType;
+            elementTypes[customType.getRuleIndex()] = customType;
```

Comment:
Maybe, the wording is not perfect, but they are _meant_ to override the auto-generated ones, because the latter are generated from the whole set of rule names.

The rule indices are coming from the generated Parser, I assume, like here:
https://github.com/protostuff/protobuf-jetbrains-plugin/blob/master/src/main/java/io/protostuff/jetbrains/plugin/ProtoParserDefinition.java#L311